### PR TITLE
Add CombineVariants, AnnotateVariants from GATK

### PR DESCRIPTION
**Pull request by ihodes**
WIP: need to figure out how to best acquire the reference genome for use by these tools…

also 
1. depends on all the VCFs passed into combine_variants
2. make annotate_variants easier to use (right now you really need to write most of what it does into the params/config in strings); this is a confusing tool

---
This change is [Reviewable](https://reviewable.io/reviews/hammerlab/biokepi/292)

### CODE REVIEW

File: src/bfx_tools/gatk.ml [hunk 1]
```diff
@@ -639,3 +647,95 @@ let mutect2
...
+    ?(configuration=Configuration.Combine_variants.default)
+    ?(edges=[])
+    ~run_with
+    ~variant_sources
+    ~output_vcf_path
+    ~genotype_merge_option
+  =
+  let open KEDSL in
+  let reference =
+    Machine.get_reference_genome run_with "" in (* how to get the reference we want to use? *)
```
Comment (smondet) [inline src/bfx_tools/gatk.ml]: pass a `~reference_genome` argument?
Like here: https://github.com/hammerlab/biokepi/blob/c2aac7c1702375485268ae2c1c74f5d9d11a53ff/src/pipeline_edsl/semantics.ml#L188


Later, when this one and #268 are merged, we can create a proper `vcf` workflow-product that (like we do in `bam_file`) carries around its reference-genome.

**Comment by ihodes** [comment]
@smondet back to you

File: src/bfx_tools/gatk.ml [hunk 2]
```diff
@@ -639,3 +647,96 @@ let mutect2
...
+  let gatk = Machine.get_tool run_with Machine.Tool.Default.gatk in
+  let reference_fasta = Reference_genome.fasta reference_genome in
+  let name = sprintf "AnnotateVariants to %s" (Filename.basename output_vcf_path) in
+  let make =
+    Machine.run_big_program run_with ~name
+      ~self_ids:["gatk"; "annotate-variants"]
+      Program.(
+        Machine.Tool.(init gatk)
+        && shf "cd %s" (Filename.dirname output_vcf_path)
+        && call_gatk ~analysis:"CombineVariants" ([
```
Comment (smondet) [inline src/bfx_tools/gatk.ml]: `CombineVariants`?

File: src/bfx_tools/gatk.ml [hunk 3]
```diff
@@ -639,3 +647,96 @@ let mutect2
...
+    ~edges:(edges @ [
+        depends_on Machine.Tool.(ensure gatk);
+        depends_on reference_fasta;
+        on_failure_activate (Remove.file ~run_with output_vcf_path);
+      ])
+
+(** Annotate variant calls with context information.
+
+    Virtually nothing is wrapped in this call; the input files (via the
+    configuration) and edges will need to be manually passed.
```
Comment (smondet) [inline src/bfx_tools/gatk.ml]: I don't understand this, how do you give the input nodes then?

Comment (ihodes) [inline src/bfx_tools/gatk.ml]: You can depend on them via `?edges`, and give the paths via the Configuration record "parameters". The spec of VariantAnnotator is a little much, so I just thought I'd get this in & add to it as I learn what the various modes do/how to use them.

File: src/bfx_tools/gatk.ml [hunk 4]
```diff
@@ -639,3 +647,96 @@ let mutect2
...
+      ~f:(function
+        | `Vcf vcf -> sprintf "--variant %s" vcf#product#path
+        | `Vcf_named_sample (sample_name, vcf) ->
+          sprintf "--variant:%s %s" sample_name vcf#product#path) in
+  let merge_option =
+    "-genotypeMergeOptions " ^
+    match genotype_merge_option with
+    | `Uniquify -> "UNIQUIFY"
+    | `Prioritize samples ->
+      "PRIORITIZE --priority " ^ String.concat ~sep:"," samples
```
Comment (iskandr) [inline src/bfx_tools/gatk.ml]: The priority list isn't required:

>UNION: assumes that each ROD source represents the same set of samples (although this is not enforced). It uses the priority list (if provided) to emit a single record instance at every position represented in the input RODs.

(though I'm not sure what GATK does without it)

Also, shouldn't that be just one dash in `-priority`? 

(In the GATK example it's `-genotypeMergeOptions PRIORITIZE -priority foo,bar`)

Comment (ihodes) [inline src/bfx_tools/gatk.ml]: > The priority list isn't required … (though I'm not sure what GATK does without it)

Me neither, and later in the docs it sounds as though it *is* required, or undefined behavior if it's missing… so I thought I'd just require it here.

> Also, shouldn't that be just one dash in -priority?

Good catch, thank you! GATK requires -- in some places and - in others… lots of fun.

**Pull request closed without merging**